design-proposals/vm-pool.md

```diff
@@ -0,0 +1,194 @@
+			* **Oldest** - Pick oldest VMs by creation time for scale-in first
+		* **ScaleInStatePreservation** - (Optional) specifies if and how to preserve state of VMs selected for scale-in.
+			* **Disabled** - (Default) all state for VMs selected for scale-in will be deleted
+			* **Offline** - PVCs for VMs selected for scale-in will be preserved and reused on scale-out (decreases provisioning time during scale out)
+			* **Online** - [NOTE we can't implement this until we have the ability to suspend VM memory state to a PVC] PVCs and memory for VMs selected for scale-in will be preserved and reused on scale-out (decreases provisioning and boot time during scale out)
+Each VM’s PVCs are preserved for future scale out
+* **AutohealingStrategy** - (Optional) 
+	* **None (Default)** - VM state is preserved, even during crashloop. No action is taken to auto recover the VM.
+	* **ReprovisionOnFailure** - VM is completely reprovisioned with persistent state refresh if the VM’s VMI terminates unexpectedly with VMI.Status.Phase=failed. This is useful in instances where a VM has data corruption and continues to fail liveness checks.
+
+
+## VMPool API Examples
+
+**Automatic rolling updates and "Shutdown" as scale-in strategy to optimization of boot times during scale-out**
+
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachinePool                                             
+metadata:                                                            
+  name: my-vm-pool                                                   
+spec:                                                                
+  virtualMachineRef:                                                                
+    name: my-template
+    kind: VirtualMachineConfig
+  replicas: 100                                                   
+  maxUnavailable: 10                                             
+  disruptionStrategy:
+    proactive:
+      scaleInStatePreservation: Offline
+```
+
+**Manual rolling updates and Manual scale-in strategy**
+
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachinePool                                             
+metadata:                                                            
+  name: my-vm-pool                                                   
+spec:                                                                
+  VirtualMachineConfigRef:                                                                
+    name: my-template                                                
+  replica: 100                                                   
+  disruptionStrategy:
+    unmanaged: {}
+```
+
+# Special Topics
+
+The topics in this section tackles in detail how specific use cases and functionality are handled with the VMPool abstraction.
+
+## Manually Detaching VM from VMPool
+
+A VM in a VMPool can be detached from a VMPool by removing the owner reference. This removes that VM from being actively managed by the VMPool.
```

Comment:
These operations could deserve convenience subresources on the Pool API.
And for the future it could also be desirable to instead of just detaching which is kind of like claiming, actually removing a VM from the pool by (live-)migrating it into a cloned VM with a new name.
That would let the Pool refill the name if it is certain a VM never goes back into the pool.

Reply:
> These operations could deserve convenience subresources on the Pool API.

that could make sense

> And for the future it could also be desirable to instead of just detaching which is kind of like claiming, actually removing a VM from the pool by (live-)migrating it into a cloned VM with a new name.
That would let the Pool refill the name if it is certain a VM never goes back into the pool.

yup, there wouldn't be anything preventing someone from cloning a VM managed by a VM pool.